WNPRC_EHR/src/org/labkey/wnprc_ehr/email_templates/notifications/PregnancyNotification.jsp

```diff
@@ -0,0 +1,92 @@
+<%@ page import="org.labkey.api.view.ActionURL" %>
+<%@ page import="org.labkey.api.util.Path" %>
+<%@ page import="org.labkey.dbutils.api.SimpleQueryFactory" %>
+<%@ page import="org.json.JSONArray" %>
+<%@ page import="org.labkey.api.data.CompareType" %>
+<%@ page import="org.json.JSONObject" %>
+<%@ page import="org.labkey.dbutils.api.SimplerFilter" %>
+<%@ page import="org.labkey.wnprc_ehr.TriggerScriptHelper" %>
+<%@ page import="org.labkey.wnprc_ehr.notification.PregnancyNotification" %>
+<%@ page extends="org.labkey.api.jsp.JspBase" %>
+<%
+    JSONObject params = (JSONObject) getModelBean();
+    SimpleQueryFactory queryFactory = new SimpleQueryFactory(getUser(), getContainer());
+
+    // Define a default when we don't have a value for a field
+    String NONE_SPECIFIED = "<em>Not Specified</em>";
+
+    // Grab the Id
+    String Id = params.getString(PregnancyNotification.idParamName);
+    String objectid = params.getString(PregnancyNotification.objectidsParamName);
+
+    SimplerFilter idFilter = new SimplerFilter("Id", CompareType.EQUAL, Id);
+    SimplerFilter objectidFilter = new SimplerFilter("objectid", CompareType.EQUAL, objectid);
+
+    JSONArray demographicsRows = queryFactory.selectRows("study", "demographics", idFilter);
+
+    JSONObject demographicsInfo = demographicsRows.getJSONObject(0);
+
+    //prepaid = demographicsInfo.getString("prepaid");
+    String gender  = demographicsInfo.getString("gender_fs_meaning");
+
+    // Construct the href for the animal abstract link.
+    String hrefForAnimalAbstract = (new Path(ActionURL.getBaseServerURL(), "ehr", getContainer().getPath(), "animalHistory.view")).toString();
+    hrefForAnimalAbstract += "?#subjects:" + Id + "&inputType:singleSubject&showReport:1&activeReport:abstract";
+%>
+
+<style type="text/css">
+    table {
+        margin-left: 10px;
+        border: 1px solid black;
+    }
+    tr > td:first-child {
+        font-weight: bold;
+        padding-right: 8px;
+        text-align: right;
+    }
+</style>
+
+<p>
+    <%= h(Id) %> has been reported as pregnant.  More details can be found below:
+</p>
+<h3>Demographic Information</h3>
+<table>
+    <tr>
+        <td>Animal ID:</td>
+        <td>
+            <a href="<%= hrefForAnimalAbstract %>">
+                <%= h(Id) %>
+            </a>
+        </td>
+    </tr>
+    <tr>
+        <td>Sex:</td>
+        <td><%= gender %></td>
+    </tr>
+</table>
+
+<h3>Pregnancy Information</h3>
+<%
+    JSONArray pregnancies = queryFactory.selectRows("study", "pregnancies", objectidFilter);
+    JSONObject pregnancy = pregnancies.length() > 0 ? pregnancies.toJSONObjectArray()[0] : new JSONObject();
+
+    String dateOfConception = pregnancy.getString("date_conception");
+    String dateDue = pregnancy.getString("date_due");
+    String sireid = pregnancy.getString("sireid");
+%>
+
+<table>
+    <tr>
+        <td>Date of Conception:</td>
+        <td><%= pregnancy.optString("date_conception", NONE_SPECIFIED) %></td>
```

Comment:
I am trying to resolve this problem since apparently in the near future this will no longer work as expected. I found this documentation on the LabKey portal: 
https://www.labkey.org/bb720087-017a-1037-be03-4d689e5f35a0/announcements-thread.view?entityId=a9578657-6566-1037-b690-f899c3d9e558&_docid=thread%3Aa9578657-6566-1037-b690-f899c3d9e558

I am not 100% sure how to encode the JSON Object with I am guessing I should use the unsafe method of your HtmlString.java class? Something like this:

    HtmlString dateOfConception = unsafe(pregnancy.getString("date_conception"));
    HtmlString dateDue = unsafe(pregnancy.getString("date_due"));
    HtmlString sireid = unsafe(pregnancy.getString("sireid"));

Reply:
unsafe() passes through the raw value without encoding - that's why it's called unsafe, because it might introduce XSS from non-encoded HTML.

You can use <%= h(pregnancy.getString("sireid") %> which will HTML-encode the raw string value.

Reply:
Would this work, I realized I sent you the code that is used by the JSP to generate the HTML text, would this work than:

<table>
    <tr>
        <td>Date of Conception:</td>
        <td><%= h(pregnancy.optString("date_conception", NONE_SPECIFIED)) %></td>
    </tr>
    <tr>
        <td>Estimated Due Date:</td>
        <td><%= h(pregnancy.optString("date_due", NONE_SPECIFIED)) %></td>
    </tr>
    <tr>
        <td>Sire:</td>
        <td><%= h(pregnancy.optString("sireid", NONE_SPECIFIED)) %></td>
    </tr>
</table>

Reply:
Yes, that looks good